Memory_Management/Heap_Allocation.md

```diff
@@ -1,91 +1,470 @@
+
+And that's it
+
+### Heap Initialization
+
+The heap initialization depends on how the operating system is implemented, if for example we are creating a kernel without multitasking we will need just to create a single heap, and initialize it probably just once,  but if we are implementing a multitasking os, with paging enabled, and every process with it's own addressing space, we will need to create a new heap and initialize it for every single process that the os is going to create (and we should remember to deallocate it once the process is dread). 
+
+The good news is that in both scenarios the initialization is more or less the same, and the only difference is that we have a single heap in one case and multiple heaps in the other (if you are arrived here you know that we are talking about virtual memory allocation, and that the address space is much bigger than the available memory). 
+
+What is really needed to initialize the Heap is to decide an initial size (for example 8k), and create a single node: 
+
+```c
+
+uint64_t *heap_start;
+uint64_t *heap_end;
+void initialize_heap() {
+  heap_start = INITIAL_HEAP_ADDRESS //Remember is a virtual address;
+  heap_start->next = NULL;
+  heap_start->next = NULL; // Just make sure that prev and next are not going anywhere
+  heap_start->status = free;
+  heap_start->size = INITIAL_HEAP_SIZE // 8096 = 8k;
+  heap_end = heap_start
+} 
+```
+
+Initially the head and the tail of the list are the same since we have just a single node.
+
+And that's it, that is how the heap is initialized, with a single node. So the first allocation will trigger a split from that node... and so on...
+
+### Heap Expansion
+
+One final part that we will explain briefly, is what happens when we reach the end of the heap, so imagine the following scenario we have done a lot of allocations, most of the heap nodes are used and the few nodes are pretty small, and the next allocation request will not find a suitable node because the requested size is bigger than any free node available. So the allocator searched through the heap, and reached the end without success. What happens next?
+
+Here is where the Virtual Memory Manager will join the game. Basically what happens is:
 
-Of course things could have been different if we assumed that the addressed returned by the allocator were at a page boundary (i.e. *a* was starting at 0x10985 and *b* at 0x10990, this would have caused to have *c* be outside the current page boundary, so it would have caused a page fault, but we will talk more about it later). 
+* The Heap Allocator first check if we have reached the end of the address space available (most unlikely) 
+* If not it will ask to the virtual Memory Manager to map one or more (depdending on implementation choiice) phyisical page(s) at the address(es) starting from `heap_end + heap_end->size + sizeof(heap_node)`
```

Comment:
lots of brackets, perhaps change it to something like: "If not it will ask the VMM to map a number of pages (depending on the implementation) at the address `heap_end + heap_end->size + sizeof(heap_node)`".